### DISCUSSION TITLE
Using check_service to tell when a service is not installed

### DISCUSSION THREAD

**Discussion by ghost** [ConSol-Monitoring/snclient#80] [General]
I'm testing SNClient at the moment and everything seems to be going well, but I have run into a bit of an issue with the check_service command and I'm not sure if this is intended or not. 

If I'm using the check_process command, I can "upgrade" the result from unknown to critical if a process is not running, for example:

```
:/usr/local/nagios/libexec$ ./check_nrpe -H 172.31.100.1 -c check_process -a "process=winword"
UNKNOWN - check_process failed to find anything with this filter. |'count'=0;;;0 'rss'=0B;;;0 'virtual'=0B;;;0 'cpu'=0%;;;0
```

Now, if I add empty-syntax and empty-state, I can manually change this:

```
:/usr/local/nagios/libexec$ ./check_nrpe -H 172.31.100.1 -c check_process -a "process=winword" "empty-syntax=CRITICAL: Word not running" "empty-state=2"
CRITICAL: Word not running |'count'=0;;;0 'rss'=0B;;;0 'virtual'=0B;;;0 'cpu'=0%;;;0
```

Obviously, we wouldn't need to check if Word is running but this demonstrates the point. With check_service however, I cannot do the same thing:

```
:/usr/local/nagios/libexec$ ./check_nrpe -H 172.31.100.1 -c check_service -a "service=madeupservice"
UNKNOWN - failed to open service madeupservice: failed to find service name madeupservice: The specified service does not exist as an installed service.
```

Using the same empty-syntax and empty-state values:

```
:/usr/local/nagios/libexec$ ./check_nrpe -H 172.31.100.1 -c check_service -a "service=madeupservice" "empty-syntax=CRITICAL: Service not installed" "empty-state=2"
UNKNOWN - failed to open service madeupservice: failed to find service name madeupservice: The specified service does not exist as an installed service.
```

Is this intended functionality or could this be altered at all? The idea is that we would want to make sure we have a particular service installed on machines and this would be an easy way to check.

**Comment by sni** [comment]
One of the reasons is that it is way easier to retrieve the process table than the services. So in case of the check_process all processes are fetched and the process= argument is used as filter.
If nothing matches, the empty clause will be used.
In case of services, when there is a service argument, then snclient directly tries to open only this specific service and in case this doesn't work, it is treated as error, not as empty list.
Thinking about it, i don't think there is a generic solution for this, at least no easy one.
One idea might be to use the error message in the empty-message instead of bailing out. Then the
empty-message and the empty-state could be used and the original error message could be made available as macro.
